### PR TITLE
Data Binding

### PR DESCRIPTION
**Pull request by kedzie**
resolves simpligility/android-maven-plugin#683

There is a bug, and possible cleanup required.  The bug is very strange.  The integration test project has a library and an apk.  both work independently.  But they don't work from the parent project.  The problem has to do with where you invoke maven from.

So this works:
databinding/library >  mvn clean install

This does not:
databinding > mvn -f library clean install

It tries to process layouts from the app.  It is as if the AnnotationProcessor is scanning to filesystem for layouts to process, which doesn't make much sense to me.

### CODE REVIEW

**Comment by kedzie** [comment]
Running maven from a different folder works, as long as there isn't any data binding generated classes under the folder you are running from....  I don't know why the annotation processor's behavior would depend on where it is invoked from...

**Comment by kedzie** [comment]
@mosabua  any thoughts?

**Comment by mosabua** [comment]
I have to catch up on merges.. will do in the next days.

**Comment by mosabua** [comment]
Whoa.. those are some big changes.. could you potentially rebase to master and see if it all works? I am working through the other PRs as well.

File: pom.xml [hunk 1]
```diff
@@ -334,9 +346,22 @@
...
+        <groupId>com.android.databinding</groupId>
+        <artifactId>dataBinder</artifactId>
+        <version>1.0-rc2</version>
+      </dependency>
+
     </dependencies>
   </dependencyManagement>
 
+  <repositories>
+    <repository>
```
Comment (mosabua) [inline pom.xml]: We can not have a repository element in the pom if we deloy to Central. Why is this needed anyway?

File: src/main/resources/META-INF/m2e/lifecycle-mapping-metadata.xml [hunk 2]
```diff
@@ -0,0 +1,48 @@
+<!--
```
Comment (mosabua) [inline src/main/resources/META-INF/m2e/lifecycle-mapping-metadata.xml]: Apache license required..

File: src/test/projects/databinding/library/src/main/AndroidManifest.xml [hunk 3]
```diff
@@ -0,0 +1,9 @@
+<?xml version="1.0" encoding="utf-8"?>
+<manifest xmlns:android="http://schemas.android.com/apk/res/android"
+    package="com.simpligility.android.databinding.library">
+
+    <uses-permission android:name="android.permission.CALL_PHONE" />
```
Comment (mosabua) [inline src/test/projects/databinding/library/src/main/AndroidManifest.xml]: is this really required... its just an example app..

File: src/test/projects/databinding/pom.xml [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+                  <dataBinding>
+                     <skip>false</skip>
+                  </dataBinding>
+               </configuration>
+            </plugin>
+         </plugins>
+      </pluginManagement>
+   </build>
+
+   <repositories>
```
Comment (mosabua) [inline src/test/projects/databinding/pom.xml]: again. cant have a repository required

File: src/test/projects/databinding/pom.xml [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+            <version>23.0.1</version>
+         </dependency>
+      </dependencies>
+   </dependencyManagement>
+
+   <build>
+      <plugins>
+         <plugin>
+            <artifactId>maven-compiler-plugin</artifactId>
+            <version>2.3.2</version>
```
Comment (mosabua) [inline src/test/projects/databinding/pom.xml]: thats a pretty old version.

**Comment by mosabua** [comment]
I pulled this into the databinding branch in this project now and I am in the process of fixing it there. I upgraded dependencies so far and will do more. Ideally you could help with pull requests for that branch..

File: pom.xml (hunk 1 above)
Comment (mosabua) [inline pom.xml]: Removed in branch. Its not needed.

File: src/test/projects/databinding/pom.xml [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+         <dependency>
+            <groupId>com.android.support</groupId>
+            <artifactId>design</artifactId>
+            <version>23.0.1</version>
+            <type>aar</type>
+         </dependency>
+         <dependency>
+            <groupId>com.android.support</groupId>
+            <artifactId>support-annotations</artifactId>
+            <version>23.0.1</version>
```
Comment (mosabua) [inline src/test/projects/databinding/pom.xml]: Should be using latest version..

Comment (kedzie) [inline src/test/projects/databinding/pom.xml]: great. the biggest issue is that the annotation processor looks through the
filesystem for some files, not specified folders.  result is that build
always needs to be run from the root of module, I suppose Gradle might
guarantee that.. well maven doesn't.  I'm not even sure how to fix it and I
haven't found the code for the actual annotation processor.  I have aosp
source pulled I need to just search for it there..

Comment (mosabua) [inline src/test/projects/databinding/pom.xml]: Maybe we should just upgrade the tools and stuff and NOT worry about the databinding for this release. Wdyt?

Comment (mosabua) [inline src/test/projects/databinding/pom.xml]: I am going to do that now... in a separate branch and pull request.

**Comment by mosabua** [comment]
The SDK upgrade is now done as a separate PR and merged via https://github.com/simpligility/android-maven-plugin/commit/de4075d1c341163422191579836259bf52abb398

Lets rebase and fix this up for the next release.

**Comment by mosabua** [comment]
Hi @kedzie 

I released 4.4.1 now and would like to get this into a soon to follow release. Could you rebase and sort out the issues with the dependencies of the example project?

**Comment by kedzie** [comment]
yes I am trying to locate the code for the core annotation processor, which is google code and has a major issue which I don't know is possible to fix from the "outside".  So I fear the only way is to fork the google annotation processor and fix it.  I am going to devote some time today searching for the code.  I am going to pull all the AOSP source and grep for the classname. 

 I can't imagine why they would write code so stupid.  The processor generates some binary files _(with a ```.bin``` extension)_ during execution.  Then later phases in the processor locates and reads those files.  And to read them it will blindly search the filesystem recursively from the current folder _(the process's current folder)_.   So if you are building an aggregator project with multiple modules, which use data binding, the processor will pick up the binary files from all those modules.  If you run the build from a different folder altogether with ```mvn -f ../somefolder/pom.xml install```, it won't find any at all.  So incredibly stupid... 

The solution seems simply, the annotation processor needs to know the output folder of the project and limit the search to there..  which I imagine cannot be too difficult.  after all you can pass parameters to annotation processors using property files in the META-INF folder.. and those files could be generated by the plugin..

**Comment by mosabua** [comment]
Maybe post about this dilemma on the SDK dev list and ask for advice. This seems a crazy approach to support databinding..

**Comment by kedzie** [comment]
Well if we locate the code we can always just submit a PR to [AOSP](https://android-review.googlesource.com/).  That would be the proper solution.

**Comment by kedzie** [comment]
I believe the class we need to find code for is ```android/databinding/annotationprocessor/ProcessDataBinding.class```.  If you care to help.

**Comment by mosabua** [comment]
I wont be able to help.. too much other stuff. I would still suggest to ask on the ADT dev list

**Comment by kedzie** [comment]
Which list exactly.  You have a link/address?

**Comment by mosabua** [comment]
https://groups.google.com/forum/#!forum/adt-dev  this is where the sdk dev team lives..

**Comment by kedzie** [comment]
I found the [Data-Binding source](https://android.googlesource.com/platform/frameworks/data-binding)

**Comment by mosabua** [comment]
@kedzie if you still want this .. can you rebase off master? Otherwise I might close this.

**Comment by kedzie** [comment]
yeah I can. just been busy with other stuff.  I think I did post something
on the dev list.

**Comment by mosabua** [comment]
Too old .. not going to worry unless this is reopened or replaced with a new PR.
